### ISSUE TITLE
Support for agents?

### ISSUE THREAD

**Issue by jglick** [Azure/azure-artifact-manager-plugin#11]
[This code](https://github.com/Azure/azure-artifact-manager-plugin/blob/53373e8cb691d47d223b71b8d740d5623c041681/src/main/java/com/microsoft/jenkins/artifactmanager/AzureArtifactManager.java#L81-L91) does not look like it would work at all when run on an agent. I do not see any indication that there is test coverage for this either. See `artifact-manager-s3` for what you need to do in order to make `achiveArtifacts`, `stash`, `unstash`, and `unarchive` work with agents.

**Comment by gavinfish** [comment]
That code just reuses some methods from [azure storage plugin](https://github.com/jenkinsci/windows-azure-storage-plugin) which depends on [azure storage sdk](https://github.com/Azure/azure-storage-java) to upload/download. It can work on an agent and I will add more test codes for this plugin.

**Comment by jglick** [comment]
Sorry, I missed that this was making a call to another Jenkins plugin. But based on [this](https://github.com/jenkinsci/windows-azure-storage-plugin/blob/dea976f071bc2dc50719a4eb1b2135101828b56f/src/main/java/com/microsoftopentechnologies/windowsazurestorage/service/UploadService.java#L229), it does so by transferring the file contents back to the master via the Remoting channel before uploading them to Azure, defeating one of the [key justifications for JEP-202](https://github.com/jenkinsci/jep/blob/master/jep/202/README.adoc#motivation). That is why `artifact-manager-s3` uses presigned URLs.

**Comment by gavinfish** [comment]
That's right. So what I should do here is to make uploading/downloading threads extend `MasterToSlaveFileCallable`, right?

**Comment by jglick** [comment]
Yes, you can `act` on that to do a chunk of work on the agent side. Making the security reasonable is the tricky part, hence the decision to use presigned URLs in the JEP-202 reference implementation: only the master need ever have any cloud credentials.

**Comment by gavinfish** [comment]
Thanks @jglick , I will figure this out, maybe I can use Azure Active Directory token lifetime  which is also mentioned in JEP-202.
